### ISSUE TITLE
[virtio-serial] qemu exit after hot plug vcpu on Win2016

### ISSUE THREAD

**Issue by zyb521** [virtio-win/kvm-guest-drivers-windows#195]
Hi everyone,

We found that in Windows server 2016, after the hot plug vcpu，qemu process directly exit some times， I'm not sure if it affects other Windows systems.

By preliminary analysis, is determined to be caused by the virtio-serial driver, our current code is based on the 0.1.130 version and incorporated into the part of the patch, by gradually backing out the patch, it is determined to "virtiolib: Rewrite VirtIORing.c, 9b89619f7787088dd026da01e3fb566ccb9b9397" commit.
we are throwing the specific code location of the problem.

There was also a problem with the community's last master code, and the Fedora 0.1.141 did not include the commit, so it did not find this problems.

Please check this issue.
tks all.

some information:
Hostos: 3.10.0-327.61.59.57_15.x86_64 is based on CentOS 7.2
qemu: 2.6
qemu exit log:
![image](https://user-images.githubusercontent.com/5163063/32988302-2dd02d3c-cd3c-11e7-8a57-2a8b16bb9c7d.png)

**Comment by zyb521** [comment]
qemu cmd:
`/usr/bin/qemu-kvm -name pvc_win2016_64_standard.qcow2,debug-threads=on -S -object secret,id=masterKey0,format=raw,file=/var/run/libvirt/qemu/domain-362-pvc_win2016_64_stand/master-key.aes -machine pc-i440fx-2.6,accel=kvm,usb=off -cpu host,hv_time,+kvm_pv_eoi,hv_relaxed,hv_vapic,hv_spinlocks=0x1fff -m 8192 -realtime mlock=off -smp 2,maxcpus=8,sockets=8,cores=1,threads=1 -uuid 9cbe73d4-0ad9-4726-9ef9-045ccbf65109 -no-user-config -nodefaults -chardev socket,id=charmonitor,path=/var/run/libvirt/qemu/domain-362-pvc_win2016_64_stand/monitor.sock,server,nowait -mon chardev=charmonitor,id=monitor,mode=control -rtc base=utc -no-shutdown -boot strict=on -device pci-bridge,chassis_nr=1,id=pci.1,bus=pci.0,addr=0x7 -device pci-bridge,chassis_nr=1,id=pci.2,bus=pci.0,addr=0x8 -device pci-bridge,chassis_nr=1,id=pci.3,bus=pci.0,addr=0x9 -device pci-bridge,chassis_nr=1,id=pci.4,bus=pci.0,addr=0xa -device pci-bridge,chassis_nr=1,id=pci.5,bus=pci.0,addr=0xb -device pci-bridge,chassis_nr=1,id=pci.6,bus=pci.0,addr=0xc -device pci-bridge,chassis_nr=1,id=pci.7,bus=pci.0,addr=0xd -device pci-bridge,chassis_nr=1,id=pci.8,bus=pci.0,addr=0xe -device piix3-usb-uhci,id=usb,bus=pci.0,addr=0x1.0x2 -device virtio-scsi-pci,id=scsi0,bus=pci.0,addr=0x4 -device virtio-serial-pci,id=virtio-serial0,bus=pci.0,addr=0x19 -drive file=/Images/TestImg/pvc_win2016_64_standard.qcow2,format=qcow2,if=none,id=drive-scsi0-0-0-0,cache=none,aio=threads -device scsi-hd,bus=scsi0.0,channel=0,scsi-id=0,lun=0,drive=drive-scsi0-0-0-0,id=scsi0-0-0-0,bootindex=1 -netdev tap,fd=36,id=hostnet0,vhost=on,vhostfd=38 -device virtio-net-pci,netdev=hostnet0,id=net0,mac=10:07:10:08:36:71,bus=pci.0,addr=0x3 -chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 -chardev socket,id=charchannel0,path=/var/run/libvirt/qemu/pvc_win2016_64_standard.qcow2.extend,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=1,chardev=charchannel0,id=channel0,name=org.qemu.guest_agent.1 -chardev socket,id=charchannel1,path=/var/run/libvirt/qemu/pvc_win2016_64_standard.qcow2.agent,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=2,chardev=charchannel1,id=channel1,name=org.qemu.guest_agent.0 -chardev socket,id=charchannel2,path=/var/run/libvirt/qemu/pvc_win2016_64_standard.qcow2.hostd,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=3,chardev=charchannel2,id=channel2,name=org.qemu.guest_agent.2 -chardev socket,id=charchannel3,path=/var/run/libvirt/qemu/pvc_win2016_64_standard.qcow2.upgraded,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=4,chardev=charchannel3,id=channel3,name=org.qemu.guest_agent.3 -vnc 0.0.0.0:0 -device cirrus-vga,id=video0,vgamem_mb=8,bus=pci.0,addr=0x2 -device virtio-balloon-pci,id=balloon0,bus=pci.0,addr=0x5 -device isa-debugcon,iobase=0x402,chardev=seabios -chardev file,id=seabios,path=/var/log/libvirt/qemu/pvc_win2016_64_standard.qcow2.seabios,mux=off,append=on -msg timestamp=on`

**Comment by ladipro** [comment]
Hello and thank you for reporting this.

> There was also a problem with the community's last master code, and the Fedora 0.1.141 did not include the commit, so it did not find this problems.

Just to make sure I understand - you were able to reproduce the problem with vanilla virtio-win/master drivers without your changes, is that correct?

> some information:
> Hostos: 3.10.0-327.61.59.57_15.x86_64 is based on CentOS 7.2
> qemu: 2.6

Is this QEMU 2.6 as shipped in CentOS or does it have any modifications?

Thank you!
Ladi

**Comment by ladipro** [comment]
I tried reproducing this and I'm getting a consistent QEMU crash with Windows Server 2016 VM after a vCPU hot-plug. I'm on a new clean install with no virtio-win drivers installed. It's not a BSOD and no event is triggered over a remote debugger connection. The QEMU process just exits.

So this definitely deserves further investigation although it may not be the same thing you're seeing.

Does hot-plug work for you if you disable the virtio-serial driver?

**Comment by YanVugenfirer** [comment]
Similar issue: https://bugzilla.redhat.com/show_bug.cgi?id=1445603

**Comment by zyb521** [comment]
@ladipro 
Yes，Just like you found it.
It is not a BSOD, Qemu process just exits after hot-plug vCPU.

If disable the virtio-serial driver, That's OK.
I guess because of virtio-serial using vcpu number in some code, like a wdf bug which workaround by commit 3628672ef8beee9aa22529f2dc149609bfab810f

And Windows 2012 R2 meet the same issue.

**Comment by zyb521** [comment]
> I'm on a new clean install with no virtio-win drivers installed. It's not a BSOD and no event is triggered over a remote debugger connection. The QEMU process just exits.
 Is vm‘s memory less than 2G in your test？
My colleagues told me, Virtio driver not installed is verified many times (every day running test), No one time found any problem.

**Comment by ladipro** [comment]
Spot on! My VM had only 2G of memory so I was running into the Bugzilla that Yan linked. That's a suspected Windows bug and unrelated to virtio-win.

With >2G of memory I can actually reproduce a virtio-related QEMU crash which has very likely the same root cause as your crash.

`(qemu) qemu-system-x86_64: Guest moved used index from 253 to 176`

Looking into it now.

**Comment by zyb521** [comment]
Yes，qemu process exit with log "Looped descriptor";
It can be found in function “virtqueue_get_avail_bytes” and "virtqueue_pop";
virtio driver give a wrong des num.

**Comment by ladipro** [comment]
I believe that the bug is now understood.

First off, it is not a regression. The reason you're hitting it only with the rewritten virtio ring code is likely just a slightly different timing.

Here's a minimal repro:

Run QEMU with:
```
  -chardev socket,host=localhost,port=4467,server,nowait,id=ser \
  -device virtserialport,chardev=ser,name=port1 \
```
And make sure the socket is stuffed with data:
`  $ nc localhost 4467 </dev/zero`

Then in the Windows VM simply open the port:
```
  C:\>copy con \\.\port1
  hello<enter>
```

And hot-add CPUs:

```
  (qemu) cpu-add 2
  (qemu) cpu-add 3
  ...
```
QEMU will crash with either **Looped descriptor** or **Guest moved used index from ... to ...**

Here's what happening. VIOSerialPortEvtDeviceD0Exit disables interrupts on the read queue but it doesn't stop it. The host side is still sending data. This causes VIOSerialDiscardPortDataLocked, called from VIOSerialPortEvtDeviceD0Exit, to loop extensively and potentially indefinitely. As long as the device has data to send, the loop runs, re-queuing the returned buffers.

Then, when VIOSerialDiscardPortDataLocked convinces itself that it's processed everything, VIOSerialPortEvtDeviceD0Exit proceeds to call VIOSerialDrainQueue which uses virtqueue_detach_unused_buf to essentially un-push already pushed buffers to the queue. Obviously this must not be done on a live queue. The driver is racing here with the device which is still trying to fill buffers and return them to the driver. This is where the QEMU asserts fire.

This will be fixed by making sure that virtqueue_detach_unused_buf runs after the device has been stopped. Or by eliminating virtqueue_detach_unused_buf altogether. The virtio-input driver seems to have the same issue.

---

It's worth mentioning that the serial driver (and maybe others too) are fundamentally broken when it comes to cpu hot-add, or any other event that causes resource rebalancing. The stream of data coming to and from the device is expected to be unaffected by the rebalance. The way it's implemented now, data coming from the device might be lost if they happen to be in the queue but not yet picked up by the driver when the event occurs (VIOSerial**Discard**PortDataLocked). Not sure if the other direction is fine, maybe Windows waits for writes to finish but it must be verified.

Since we likely can't avoid re-initializing the device on rebalance (BAR mappings may change), the driver needs to internally buffer, not discard, all data read from the queue and act as if they were popped from the queue after the rebalance.

**Comment by zyb521** [comment]
@ladipro 
Thanks for your quick reply.

> It's worth mentioning that the serial driver (and maybe others too) are fundamentally broken when it comes to cpu hot-add, or any other event that causes resource rebalancing. 

We also found the virtio-serial drivers have a reload process after cpu hot-add, and the serial port data transmission are not affected.
This looks like only wdf driver have this problem.

**Comment by ladipro** [comment]
Here's a stop-gap patch which fixes the crash for me. Can you please give it a try?
https://github.com/ladipro/kvm-guest-drivers-windows/commit/ffce69a8fe4cbdaaf86e3623d0c5a32dc6a07bf5

> We also found the virtio-serial drivers have a reload process after cpu hot-add, and the serial port data transmission are not affected.

If by transmission you mean writing to the virtual serial port on the guest, then I agree, it doesn't seem to be affected. If you mean any data transfer, then I disagree. I have verified that host->guest data is really lost during cpu hot-add.

**Comment by zyb521** [comment]
OK. I will try your patch.

**Comment by zyb521** [comment]
It is a nice patch. 
After multiple tests and no problem is found.

**Comment by zyb521** [comment]
Patch has been merged.
Thanks all

**Comment by ladipro** [comment]
The crash is tracked by:
https://bugzilla.redhat.com/show_bug.cgi?id=1520352
and will be fixed in virtio-win >=0.1.145

The data loss is tracked by:
https://bugzilla.redhat.com/show_bug.cgi?id=1520955
